### PR TITLE
Add or_insert_with_key to Entry of HashMap/BTreeMap

### PR DESCRIPTION
**Pull request by ChaiTRex**
Going along with `or_insert_with`, `or_insert_with_key` provides the `Entry`'s key to the lambda, avoiding the need to either clone the key or the need to reimplement this body of this method from scratch each time.

This is useful when the initial value for a map entry is derived from the key. For example, the introductory Rust book has an example Cacher struct that takes an expensive-to-compute lambda and then can, given an argument to the lambda, produce either the cached result or execute the lambda.

---

I'm fairly new to Rust, so any optimizations, corrections to types, better names, better documentation, or whatever else would be appreciated. I'd like to thank Arnavion on freenode for helping me to implement a very similar method when I found that `or_insert_with_key` was unavailable.

As a somewhat-related note, this implements https://github.com/rust-lang/rfcs/issues/1202 from 2015, so if this pull request is accepted, that should be closed.

### CODE REVIEW

File: src/liballoc/collections/btree/map.rs [hunk 1]
```diff
@@ -2378,6 +2378,34 @@ impl<'a, K: Ord, V> Entry<'a, K, V> {
         }
     }
 
+    #[unstable(feature = "or_insert_with_key", issue = "70996")]
```
Comment (Amanieu) [inline src/liballoc/collections/btree/map.rs:2381]: Can you create a proper tracking issue instead of just pointing to the PR?

**Review by Amanieu — COMMENTED** [review]
Thanks! Could you also submit a PR to [hashbrown](https://github.com/rust-lang/hashbrown)? I would like to keep the APIs in sync.

**Comment by ChaiTRex** [comment]
@Amanieu I've made a pull request to `hashbrown` at https://github.com/rust-lang/hashbrown/pull/152.

File: src/liballoc/collections/btree/map.rs (hunk 1 above)
Comment (ChaiTRex) [reply src/liballoc/collections/btree/map.rs]: @Amanieu This is corrected in commit 3.

**Pull request merged**